src/js/components/DataTable/buildState.js

```diff
@@ -248,38 +248,43 @@ export const normalizeCellProps = (props, theme) => {
         props?.[propName]?.[context] ||
         // if the propName is used without context, it applies to all contexts
         (tableContextNames.every((n) => !props?.[propName]?.[n]) &&
-          props?.[propName]) ||
+          props?.[propName]);
+      if (value !== undefined) result[context][propName] = value;
+
+      let themeValue =
         theme?.dataTable?.[context]?.[propName] ||
         theme?.table?.[context]?.[propName];
-      if (value !== undefined) result[context][propName] = value;
 
       // pinned case
       value =
         props?.[propName]?.pinned?.[context] ||
         (context === 'body' &&
           tableContextNames.every((n) => !props?.[propName]?.pinned?.[n]) &&
           props?.[propName]?.pinned) ||
-        theme?.dataTable?.pinned?.[context]?.[propName];
-      if (value !== undefined) {
-        if (
-          propName === 'background' &&
-          theme.background &&
-          value.opacity &&
-          !value.color
-        )
-          // theme context has an active background color but the
-          // theme doesn't set an explicit color, repeat the context
-          // background explicitly
-          value.color = normalizeBackgroundColor(theme);
-
-        if (context === 'body')
-          // in case we have pinned columns, store the pinned stuff in
-          // cellProps.body.pinned
-          result[context].pinned[propName] = value;
-        else if (props.pin === true || props.pin === context)
-          // this context is pinned, use the pinned value directly
-          result[context][propName] = value;
-      }
+        value;
+
+      themeValue =
+        theme?.dataTable?.pinned?.[context]?.[propName] || themeValue;
+
+      if (
+        value !== undefined &&
+        propName === 'background' &&
+        theme.background &&
+        value.opacity &&
+        !value.color
+      )
+        // theme context has an active background color but the
+        // theme doesn't set an explicit color, repeat the context
+        // background explicitly
+        value.color = normalizeBackgroundColor(theme);
```

Comment:
I know this is pre-existing, but this looks like a bad side-effect that could actually be modifying the theme. Since value could be a reference directly into the theme, I'm wondering if this is actually modifying that part of the theme for everyone. This type of side-effect can cause some really odd and hard to find bugs.